File: content/develop/use-cases/agent-memory/php/src/AgentSession.php
<?php
// Working-memory store for an agent session, backed by a Redis Hash.
//
// Each session is one Hash document at `agent:session:{thread_id}`.
// The hash holds the running scratchpad, the current goal, a rolling
// window of recent turns (serialized as a JSON list to fit in one
// field), and a few audit fields. One `HGETALL` returns the whole
// session in a single round trip on every step of the agent loop.
//
// Every write refreshes the key's TTL with `EXPIRE`, so idle sessions
// fall off without a separate cleanup job and active sessions stay
// alive as long as the agent keeps touching them. A separate
// `LongTermMemory` (see `LongTermMemory.php`) is what survives
// beyond a session's TTL.
//
// The turn window is bounded to `maxTurns` in application code; the
// hash itself doesn't grow, so the working set per thread stays
// constant regardless of how long the agent has been running.

declare(strict_types=1);

namespace Redis\AgentMemory;

use Predis\Client;

class AgentSession
{
    public const MAX_TURNS = 20;

    public readonly Client $client;
    public readonly string $keyPrefix;
    public readonly int $defaultTtlSeconds;
    public readonly int $maxTurns;

    public function __construct(
        Client $client,
        string $keyPrefix = 'agent:session:',
        int $defaultTtlSeconds = 3600,
        int $maxTurns = self::MAX_TURNS,
    ) {
        $this->client = $client;
        $this->keyPrefix = $keyPrefix;
        $this->defaultTtlSeconds = $defaultTtlSeconds;
        $this->maxTurns = $maxTurns;
    }

    public function sessionKey(string $threadId): string
    {
        return $this->keyPrefix . $threadId;
    }

    public function newThreadId(): string
    {
        return substr(bin2hex(random_bytes(6)), 0, 12);
    }

    /**
     * Create a fresh working memory for a thread. Overwrites any
     * existing session at the same key. The agent normally calls
     * this once per thread at the first turn and relies on `load` /
     * `appendTurn` for subsequent steps.
     *
     * @return array<string,mixed>
     */
    public function start(
        string $threadId,
        string $user = 'default',
        string $agent = 'default',
        string $goal = '',
        ?int $ttlSeconds = null,
    ): array {
        $ttl = $ttlSeconds ?? $this->defaultTtlSeconds;
        $now = microtime(true);
        $state = [
            'thread_id' => $threadId,
            'user' => $user,
            'agent' => $agent,
            'goal' => $goal,
            'scratchpad' => '',
            'turn_count' => 0,
            'created_ts' => $now,
            'last_active_ts' => $now,
            'recent_turns' => [],
            'ttl_seconds' => $ttl,
        ];
        $this->write($state, $ttl);
        return $state;
    }

    /**
     * Return the session state, or `null` if it has expired.
     *
     * @return array<string,mixed>|null
     */
    public function load(string $threadId): ?array
    {
        $key = $this->sessionKey($threadId);
        $raw = $this->client->hgetall($key);
        if (!$raw) {
            return null;
        }
        $ttl = (int) $this->client->ttl($key);
        $turnsBlob = $raw['recent_turns'] ?? '[]';
        $turns = json_decode($turnsBlob, true);
        if (!is_array($turns)) {
            $turns = [];
        }
        return [
            'thread_id' => $threadId,
            'user' => $raw['user'] ?? 'default',
            'agent' => $raw['agent'] ?? 'default',
            'goal' => $raw['goal'] ?? '',
            'scratchpad' => $raw['scratchpad'] ?? '',
            'turn_count' => (int) ($raw['turn_count'] ?? 0),
            'created_ts' => (float) ($raw['created_ts'] ?? 0),
            'last_active_ts' => (float) ($raw['last_active_ts'] ?? 0),
            'recent_turns' => $turns,
            'ttl_seconds' => $ttl > 0 ? $ttl : 0,
        ];
    }

    /**
     * Append a turn, bound the rolling window, refresh the TTL.
     *
     * `user` and `agent` are only consulted when the session does
     * not yet exist — they seed the auto-created session so the
     * working-memory hash matches the user the caller is operating
     * against. On an existing session they're ignored; the original
     * `start` values stand.
     *
     * Read-modify-write here is last-writer-wins on the turn list if
     * two concurrent turns reach the same thread; the demo never
     * triggers that race in practice (one browser, one turn at a
     * time) but a multi-worker agent that shares a thread id would
     * wrap this in `WATCH` / `MULTI` / `EXEC` or a Lua script that
     * does the append atomically server-side.
     *
     * @return array<string,mixed>
     */
    public function appendTurn(
        string $threadId,
        string $role,
        string $content,
        ?string $user = null,
        ?string $agent = null,
        ?int $ttlSeconds = null,
    ): array {
        $state = $this->load($threadId);
        if ($state === null) {
            $state = $this->start(
                $threadId,
                user: $user ?? 'default',
                agent: $agent ?? 'default',
                ttlSeconds: $ttlSeconds,
            );
        }
        $state['recent_turns'][] = [
            'role' => $role,
            'content' => $content,
            'ts' => microtime(true),
        ];
        if (count($state['recent_turns']) > $this->maxTurns) {
            $state['recent_turns'] = array_slice(
                $state['recent_turns'], -$this->maxTurns,
            );
        }
        $state['turn_count'] = ((int) $state['turn_count']) + 1;
        $state['last_active_ts'] = microtime(true);
        $ttl = $ttlSeconds ?? $this->defaultTtlSeconds;
        $state['ttl_seconds'] = $ttl;
        $this->write($state, $ttl);
        return $state;
    }

    /**
     * Update the agent's running scratchpad and refresh TTL.
     *
     * @return array<string,mixed>|null
     */
    public function setScratchpad(
        string $threadId,
        string $text,
        ?int $ttlSeconds = null,
    ): ?array {
        $state = $this->load($threadId);
        if ($state === null) {
            return null;
        }
        $state['scratchpad'] = $text;
        $state['last_active_ts'] = microtime(true);
        $ttl = $ttlSeconds ?? $this->defaultTtlSeconds;
        $state['ttl_seconds'] = $ttl;
        $this->write($state, $ttl);
        return $state;
    }

    /**
     * Update the goal field without touching turns or the scratchpad.
     *
     * Creates the session if it doesn't exist yet — setting a goal
     * on a fresh thread is a sensible first step in the agent loop,
     * so this method covers both the "rename the goal mid-session"
     * and the "start a thread with this goal" cases.
     *
     * @return array<string,mixed>
     */
    public function setGoal(
        string $threadId,
        string $text,
        ?string $user = null,
        ?string $agent = null,
        ?int $ttlSeconds = null,
    ): array {
        $state = $this->load($threadId);
        if ($state === null) {
            return $this->start(
                $threadId,
                user: $user ?? 'default',
                agent: $agent ?? 'default',
                goal: $text,
                ttlSeconds: $ttlSeconds,
            );
        }
        $state['goal'] = $text;
        $state['last_active_ts'] = microtime(true);
        $ttl = $ttlSeconds ?? $this->defaultTtlSeconds;
        $state['ttl_seconds'] = $ttl;
        $this->write($state, $ttl);
        return $state;
    }

    public function delete(string $threadId): bool
    {
        return ((int) $this->client->del($this->sessionKey($threadId))) > 0;
    }

    /**
     * @param array<string,mixed> $state
     */
    private function write(array $state, int $ttl): void
    {
        $key = $this->sessionKey($state['thread_id']);
        $mapping = [
            'thread_id' => $state['thread_id'],
            'user' => $state['user'],
            'agent' => $state['agent'],
            'goal' => $state['goal'],
            'scratchpad' => $state['scratchpad'],
            'turn_count' => (string) $state['turn_count'],
            'created_ts' => (string) $state['created_ts'],
            'last_active_ts' => (string) $state['last_active_ts'],
            'recent_turns' => json_encode(
                $state['recent_turns'], JSON_UNESCAPED_SLASHES,
            ),
        ];
        // Predis HSET takes positional field/value arguments, not
        // an associative array — flatten the mapping into the order
        // Redis expects on the wire.
        $hsetArgs = [$key];
        foreach ($mapping as $k => $v) {
            $hsetArgs[] = $k;
            $hsetArgs[] = $v;
        }
        // MULTI/EXEC so HSET and EXPIRE either both apply or neither
        // does. A connection drop between the two writes would
        // otherwise leave the session without a TTL.
        $this->client->transaction(
            function ($tx) use ($hsetArgs, $key, $ttl) {
                $tx->hset(...$hsetArgs);
                $tx->expire($key, $ttl);
            },
        );
    }
}

File: content/develop/use-cases/agent-memory/dotnet/AgentSession.cs
using System.Globalization;
using System.Text.Json;
using StackExchange.Redis;

namespace AgentMemoryDemo;

/// <summary>
/// Working-memory store for an agent session, backed by a Redis Hash.
/// </summary>
/// <remarks>
/// <para>Each session is one Hash document at
/// <c>agent:session:{threadId}</c>. The hash holds the running
/// scratchpad, the current goal, a rolling window of recent turns
/// (serialised as a JSON list to fit in one field), and a few audit
/// fields. One <c>HGETALL</c> returns the whole session in a single
/// round trip on every step of the agent loop.</para>
///
/// <para>Every write refreshes the key's TTL with <c>EXPIRE</c>, so
/// idle sessions fall off without a separate cleanup job and active
/// sessions stay alive as long as the agent keeps touching them. A
/// separate <see cref="LongTermMemory"/> is what survives beyond a
/// session's TTL.</para>
///
/// <para>The turn window is bounded to <see cref="MaxTurns"/> in
/// application code; the hash itself doesn't grow, so the working set
/// per thread stays constant regardless of how long the agent has
/// been running.</para>
/// </remarks>
public sealed class AgentSession
{
    // How many recent turns to keep inline on the session hash. Older
    // turns flow through the event log (AgentEventLog) and the
    // long-term memory store (LongTermMemory).
    public const int DefaultMaxTurns = 20;

    private readonly IDatabase _db;
    public string KeyPrefix { get; }
    public long DefaultTtlSeconds { get; }
    public int MaxTurns { get; }

    public AgentSession(
        IDatabase db,
        string keyPrefix = "agent:session:",
        long defaultTtlSeconds = 3600,
        int maxTurns = DefaultMaxTurns)
    {
        _db = db;
        KeyPrefix = keyPrefix;
        DefaultTtlSeconds = defaultTtlSeconds;
        MaxTurns = maxTurns;
    }

    public string SessionKey(string threadId) => KeyPrefix + threadId;

    public string NewThreadId() => Guid.NewGuid().ToString("N").Substring(0, 12);

    /// <summary>
    /// Create a fresh working memory for a thread. Overwrites any
    /// existing session at the same key. The agent normally calls
    /// this once per thread at the first turn and relies on
    /// <see cref="Load"/> / <see cref="AppendTurn"/> for subsequent
    /// steps.
    /// </summary>
    public SessionState Start(
        string threadId,
        string user = "default",
        string agentName = "default",
        string goal = "",
        long? ttlSeconds = null)
    {
        long ttl = ttlSeconds ?? DefaultTtlSeconds;
        double now = UnixSeconds();
        var state = new SessionState(
            ThreadId: threadId,
            User: user,
            Agent: agentName,
            Goal: goal,
            Scratchpad: "",
            TurnCount: 0,
            CreatedTs: now,
            LastActiveTs: now,
            RecentTurns: Array.Empty<SessionTurn>(),
            TtlSeconds: ttl);
        Write(state, ttl);
        return state;
    }

    /// <summary>
    /// Return the session state, or <c>null</c> if it has expired.
    /// </summary>
    public SessionState? Load(string threadId)
    {
        string key = SessionKey(threadId);
        var raw = _db.HashGetAll(key);
        if (raw is null || raw.Length == 0) return null;
        var fields = raw.ToDictionary(e => (string)e.Name!, e => (string)e.Value!);
        TimeSpan? ttl = _db.KeyTimeToLive(key);
        long ttlSeconds = ttl is { TotalSeconds: > 0 } v ? (long)v.TotalSeconds : 0L;
        string turnsBlob = fields.GetValueOrDefault("recent_turns") ?? "[]";
        var turns = TryDeserializeTurns(turnsBlob);
        return new SessionState(
            ThreadId: threadId,
            User: fields.GetValueOrDefault("user") ?? "default",
            Agent: fields.GetValueOrDefault("agent") ?? "default",
            Goal: fields.GetValueOrDefault("goal") ?? "",
            Scratchpad: fields.GetValueOrDefault("scratchpad") ?? "",
            TurnCount: ParseLong(fields.GetValueOrDefault("turn_count"), 0),
            CreatedTs: ParseDouble(fields.GetValueOrDefault("created_ts"), 0),
            LastActiveTs: ParseDouble(fields.GetValueOrDefault("last_active_ts"), 0),
            RecentTurns: turns,
            TtlSeconds: ttlSeconds);
    }

    /// <summary>
    /// Append a turn, bound the rolling window, refresh the TTL.
    /// </summary>
    /// <remarks>
    /// <para><paramref name="user"/> and <paramref name="agentName"/>
    /// are only consulted when the session does not yet exist — they
    /// seed the auto-created session so the working-memory hash
    /// matches the user the caller is operating against. On an
    /// existing session they're ignored; the original <c>Start</c>
    /// values stand.</para>
    ///
    /// <para>Read-modify-write here is last-writer-wins on the turn
    /// list if two concurrent turns reach the same thread; the demo
    /// never triggers that race in practice (one browser, one turn at
    /// a time) but a multi-worker agent that shares a thread id would
    /// wrap this in <c>WATCH</c> / <c>MULTI</c> / <c>EXEC</c> or a
    /// Lua script that does the append atomically server-side.</para>
    /// </remarks>
    public SessionState AppendTurn(
        string threadId,
        string role,
        string content,
        string? user = null,
        string? agentName = null,
        long? ttlSeconds = null)
    {
        var state = Load(threadId)
            ?? Start(
                threadId,
                user: user ?? "default",
                agentName: agentName ?? "default",
                ttlSeconds: ttlSeconds);

        var newTurns = state.RecentTurns.ToList();
        newTurns.Add(new SessionTurn(Role: role, Content: content, Ts: UnixSeconds()));
        if (newTurns.Count > MaxTurns)
        {
            newTurns = newTurns.GetRange(newTurns.Count - MaxTurns, MaxTurns);
        }

        long ttl = ttlSeconds ?? DefaultTtlSeconds;
        var next = state with
        {
            TurnCount = state.TurnCount + 1,
            LastActiveTs = UnixSeconds(),
            RecentTurns = newTurns,
            TtlSeconds = ttl,
        };
        Write(next, ttl);
        return next;
    }

    /// <summary>
    /// Update the agent's running scratchpad and refresh the TTL.
    /// Returns <c>null</c> when the session does not exist.
    /// </summary>
    public SessionState? SetScratchpad(string threadId, string text, long? ttlSeconds = null)
    {
        var state = Load(threadId);
        if (state is null) return null;
        long ttl = ttlSeconds ?? DefaultTtlSeconds;
        var next = state with
        {
            Scratchpad = text,
            LastActiveTs = UnixSeconds(),
            TtlSeconds = ttl,
        };
        Write(next, ttl);
        return next;
    }

    /// <summary>
    /// Update the goal field without touching turns or the scratchpad.
    /// Creates the session if it doesn't exist yet — setting a goal
    /// on a fresh thread is a sensible first step in the agent loop,
    /// so this method covers both the "rename the goal mid-session"
    /// and the "start a thread with this goal" cases.
    /// </summary>
    public SessionState SetGoal(
        string threadId,
        string text,
        string? user = null,
        string? agentName = null,
        long? ttlSeconds = null)
    {
        var state = Load(threadId);
        if (state is null)
        {
            return Start(
                threadId,
                user: user ?? "default",
                agentName: agentName ?? "default",
                goal: text,
                ttlSeconds: ttlSeconds);
        }
        long ttl = ttlSeconds ?? DefaultTtlSeconds;
        var next = state with
        {
            Goal = text,
            LastActiveTs = UnixSeconds(),
            TtlSeconds = ttl,
        };
        Write(next, ttl);
        return next;
    }

    /// <summary>Drop the session immediately. Returns <c>true</c> if it existed.</summary>
    public bool Delete(string threadId) => _db.KeyDelete(SessionKey(threadId));

    /// <summary>Return active thread ids (for the demo's thread switcher).</summary>
    public List<string> ListThreads(int limit = 100)
    {
        var out_ = new List<string>();
        // SCAN via the server-set option; this stays incremental even
        // on a database with many session keys.
        var server = _db.Multiplexer.GetServer(_db.Multiplexer.GetEndPoints().First());
        foreach (var key in server.Keys(database: _db.Database, pattern: KeyPrefix + "*", pageSize: 200))
        {
            string raw = (string)key!;
            out_.Add(raw.StartsWith(KeyPrefix) ? raw.Substring(KeyPrefix.Length) : raw);
            if (out_.Count >= limit) break;
        }
        return out_;
    }

    private void Write(SessionState state, long ttl)
    {
        string key = SessionKey(state.ThreadId);
        var entries = new HashEntry[]
        {
            new("thread_id", state.ThreadId),
            new("user", state.User),
            new("agent", state.Agent),
            new("goal", state.Goal),
            new("scratchpad", state.Scratchpad),
            new("turn_count", state.TurnCount.ToString(CultureInfo.InvariantCulture)),
            new("created_ts", state.CreatedTs.ToString("F6", CultureInfo.InvariantCulture)),
            new("last_active_ts", state.LastActiveTs.ToString("F6", CultureInfo.InvariantCulture)),
            new("recent_turns", JsonSerializer.Serialize(state.RecentTurns.Select(t => new
            {
                role = t.Role,
                content = t.Content,
                ts = t.Ts,
            }))),
        };

        // MULTI/EXEC so HSET and EXPIRE either both apply or neither
        // does. A connection drop between the two writes would
        // otherwise leave the session without a TTL. We check the
        // return value of Execute() — there's no WATCH on this
        // transaction so a false here means the server rejected the
        // batch (out of memory, OOM script kill, etc.); surface it
        // rather than letting the in-memory state drift from Redis.
        var tx = _db.CreateTransaction();
        _ = tx.HashSetAsync(key, entries);
        _ = tx.KeyExpireAsync(key, TimeSpan.FromSeconds(ttl));
        if (!tx.Execute())
        {
            throw new RedisServerException("session write MULTI/EXEC was discarded");
        }
    }

    private static IReadOnlyList<SessionTurn> TryDeserializeTurns(string blob)
    {
        try
        {
            using var doc = JsonDocument.Parse(blob);
            var result = new List<SessionTurn>(doc.RootElement.GetArrayLength());
            foreach (var el in doc.RootElement.EnumerateArray())
            {
                string role = el.TryGetProperty("role", out var r) ? r.GetString() ?? "" : "";
                string content = el.TryGetProperty("content", out var c) ? c.GetString() ?? "" : "";
                double ts = el.TryGetProperty("ts", out var t) && t.TryGetDouble(out var d) ? d : 0.0;
                result.Add(new SessionTurn(role, content, ts));
            }
            return result;
        }
        catch (JsonException)
        {
            return Array.Empty<SessionTurn>();
        }
    }

    private static double UnixSeconds()
        => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;

    private static double ParseDouble(string? value, double fallback)
        => double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out var d) ? d : fallback;

    private static long ParseLong(string? value, long fallback)
        => long.TryParse(value, NumberStyles.Integer, CultureInfo.InvariantCulture, out var l) ? l : fallback;
}
